### ISSUE TITLE
data leakage with exemplars 

### ISSUE THREAD

**Issue by zdgithub** [google-research/FLAN#19]
I found that in the experiment "4.4 Instructions with few-shot exemplars" of the FLAN paper，it said `At both training and inference time, exemplars are randomly drawn from the training set`. 
During training, some examples would be sampled as exemplars. I think this may lead to data leakage?

**Comment by TheExGenesis** [comment]
Looking at the code, I was wondering the same thing.

**Comment by shayne-longpre** [comment]
Thank you for the question!

I need to double check, but it I think you're right that a few-shot training example might occasionally appear in the few-shot exemplars its prompted with. This is slightly sub-optimal when it happens, as the answer is telegraphed, but also extremely rare, so I doubt you'd see any effects in practice, except for very small datasets.

This is different from train/test data leakage which would bias the evaluation.

**Comment by zdgithub** [comment]
Thanks for your reply!

For each example in the train set, I would sample another example (i.e. 1-shot) from the same train set as its exemplar. 
Then I did a toy experiment to verify the ratio of data leakage.  i.e.
```
import random

# N is the size of the train set
def get_duplicate(N=10000):
    a = [i for i in range(N)]

    sampled_examples = set()

    # sample an exemplar for each training example
    for i in range(N):
        sampled_examples.add(random.sample(a, 1)[0])

    print(len(sampled_examples))
    print('ratio is ', len(sampled_examples) / N)

```
I found that when N=10000, about 62% training examples would be sampled as exemplars, which may result in serious data leakage。I'm not sure if I have done something wrong. 

Besides, shall I split the whole train set into 2 parts? One part is used as the true training examples set, and the other is used as the set of exemplars? If so, how to determine the ratio of the 2 parts?

**Comment by lehougoogle** [comment]
Hi zdgithub,

Do you mean few-shot in general?
For each training example, we randomly sample other training examples as exemplars.
I *think* there's no cases where the training example also appears as an exemplar in its context.

Thanks

**Comment by zdgithub** [comment]
@lehougoogle 
Yes. I focus on the few-shot FLAN. 
I agree with you that one training example doesn't also appear as its exemplar. But I think it may be the exemplar of other training examples, thus the model would know the label information of some training examples during training.

For example, A, B and C are all examples of the train set, then we may sample B as the exemplar of A, and sample C as the exemplar of B. In this way, when training on the input example B with context C, the model has known the whole label information of B from the previous input A with B as its context. I think this will lead to data leakage, won't it?

**Comment by lehougoogle** [comment]
Hi zdgithub,

Sorry, I got your point now. I'm not sure about whether it would introduce overfitting or not. But I think it indeed is worry-some.

Thanks

**Comment by shayne-longpre** [comment]
Following up further zdgithub: this is an interesting observation! Thanks for bringing it to our attention. 

I'm not sure that it's so much a concern for me (since we know it gets strong results on held-out data) so much as an opportunity to maybe improve results further, if this behaviour is indeed detrimental. As a simple fix you could split off a small validation set for few-shot exemplars from the train sets (sort of like this: https://github.com/google-research/FLAN/blob/main/flan/v2/task_configs_v1.py#L51-L55) and see if it improves results. 

I would be very curious to see this experiment, and if successful, this simple fix would be a huge improvement to the existing example generation!
